File: leetcode/53.maximum-subarray.py
#  Tag: Array, Divide and Conquer, Dynamic Programming
#  Time: O(N)
#  Space: O(N)
#  Ref: -
#  Note: -

#  Given an integer array nums, find the subarray with the largest sum, and return its sum.
#   
#  Example 1:
#  
#  Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
#  Output: 6
#  Explanation: The subarray [4,-1,2,1] has the largest sum 6.
#  
#  Example 2:
#  
#  Input: nums = [1]
#  Output: 1
#  Explanation: The subarray [1] has the largest sum 1.
#  
#  Example 3:
#  
#  Input: nums = [5,4,-1,7,8]
#  Output: 23
#  Explanation: The subarray [5,4,-1,7,8] has the largest sum 23.
#  
#   
#  Constraints:
#  
#  1 <= nums.length <= 105
#  -104 <= nums[i] <= 104
#  
#   
#  Follow up: If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
#  

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [0 for i in range(n)]

        dp[0] = nums[0]
        for i in range(1, n):
            dp[i] = max(dp[i - 1], 0) + nums[i]

        return max(dp)


# prefix-sum
class Solution(object):
    def maxSubArray(self, nums: List[int]) -> int:
        # suppose sumTo[i] = a[0] + a[1] + ... [i - 1], sumTo(0) = 0
        # the subarray a[i:j] = sumTo[j] - sumTo[i]

        largest = float('-inf')
        smallest = 0
        sumTo = 0

        for num in nums:
            sumTo += num
            largest = max(largest, sumTo - smallest)
            smallest = min(smallest, sumTo)

        return largest

# follow-up
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        return self.helper(nums, 0, len(nums) - 1)
    
    def helper(self, nums: List[int], left: int, right: int) -> int:
        if left == right:
            return nums[left]
        
        mid = (left + right) // 2
        leftSum = float('-inf')
        curSum = 0
        # Maximum sum of subarray ending at mid and going left
        for i in range(mid, left - 1, -1):
            curSum += nums[i]
            leftSum = max(leftSum, curSum)
        
        rightSum = float('-inf')
        curSum = 0
        # Maximum sum of subarray starting at mid+1 and going right
        for i in range(mid + 1, right + 1):
            curSum += nums[i]
            rightSum = max(rightSum, curSum)
        
        return max(self.helper(nums, left, mid),
                   self.helper(nums, mid + 1, right),
                   leftSum + rightSum) # Max sum crossing midpoint

File: leetcode/53.maximum-subarray.cpp
//  Tag: Array, Divide and Conquer, Dynamic Programming
//  Time: O(N)
//  Space: O(N)
//  Ref: -
//  Note: -

//  Given an integer array nums, find the subarray with the largest sum, and return its sum.
//   
//  Example 1:
//  
//  Input: nums = [-2,1,-3,4,-1,2,1,-5,4]
//  Output: 6
//  Explanation: The subarray [4,-1,2,1] has the largest sum 6.
//  
//  Example 2:
//  
//  Input: nums = [1]
//  Output: 1
//  Explanation: The subarray [1] has the largest sum 1.
//  
//  Example 3:
//  
//  Input: nums = [5,4,-1,7,8]
//  Output: 23
//  Explanation: The subarray [5,4,-1,7,8] has the largest sum 23.
//  
//   
//  Constraints:
//  
//  1 <= nums.length <= 105
//  -104 <= nums[i] <= 104
//  
//   
//  Follow up: If you have figured out the O(n) solution, try coding another solution using the divide and conquer approach, which is more subtle.
//  

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, 0);
        dp[0] = nums[0];
        for (int i = 1; i < n; i++) {
            dp[i] = max(0, dp[i - 1]) + nums[i];
        }

        return *max_element(dp.begin(), dp.end());
    }
}; 

// prefix-sum
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        
        int sumJ = 0;
        int sumI = 0;
        int largest = INT_MIN;
        
        // sub[i+1: j] = sum[:j] - sum[:i]
        for (auto i = 0; i < nums.size(); ++i) {
            sumJ += nums[i];
            largest = max(largest, sumJ - sumI);
            sumI = min(sumI, sumJ);
        }
        
        return largest;
    }
};

// follow-up
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        return helper(nums, 0, nums.size() - 1);
    }

    int helper(vector<int>& nums, int left, int right) {
        if (left == right) {
            return nums[left];
        }

        int mid = (left + right) >> 1; 
        int cur = 0;
        int left_max = INT_MIN;
        for (int i = mid; i >= left; i--) {
            cur += nums[i];
            left_max = max(left_max, cur);
        }

        cur = 0;
        int right_max = INT_MIN;
        for (int i = mid + 1; i <= right; i++) {
            cur += nums[i];
            right_max = max(right_max, cur);
        }

        int child_max = max(helper(nums, left, mid), helper(nums, mid + 1, right));
        return max(child_max, left_max + right_max);

    }
};